src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -257,30 +260,45 @@ private void HandleRemovedDocuments(RequestContext context, DiagnosticParams[] p
             }
         }
 
-        private bool DiagnosticsAreUnchanged(Dictionary<Document, DiagnosticParams> documentToPreviousDiagnosticParams, Document document)
+        /// <summary>
+        /// Returns true if diagnostics have changed since the last request and if so,
+        /// calculates a new resultId to use for subsequent computation and caches it.
+        /// </summary>
+        /// <param name="documentToPreviousDiagnosticParams">the resultIds the client sent us.</param>
+        /// <param name="document">the document we are currently calculating results for.</param>
+        /// <param name="newResultId">the resultId to report new diagnostics with if changed.</param>
+        private bool HaveDiagnosticsChangedAndUpdateCache(
```

Comment:
up to you, but this name feels clunky.  i would just remove the 'AndUpdateCache' and allow the docs/out-param to be clear enough.